File: 0670. Maximum Swap/Solution.java
class Solution {
    public int maximumSwap(int num) {
        char[] chars = Integer.toString(num).toCharArray();
        int[] maxIndex = new int[10];
        for (int i = 0; i < chars.length; i++) {
            maxIndex[chars[i] - '0'] = i;
        }
        for (int i = 0; i < chars.length; i++) {
            int x = chars[i] - '0';
            for (int j = 9; j > x; j--) {
                if (maxIndex[j] > i) {
                    char temp = chars[i];
                    chars[i] = chars[maxIndex[j]];
                    chars[maxIndex[j]] = temp;
                    return Integer.parseInt(new String(chars));
                }
            }
        }
        return Integer.parseInt(new String(chars));
    }
}

File: 0670. Maximum Swap/Solution.cpp
#include <string>
using namespace std;

class Solution {
public:
    int maximumSwap(int num)
    {
        string s = to_string(num);
        int n = s.size(), max_idx = n - 1, a = -1, b;
        for (int i = n - 2; i >= 0; --i) {
            if (s[i] > s[max_idx])
                max_idx = i;
            else if (s[i] < s[max_idx]) {
                a = i;
                b = max_idx;
            }
        }
        if (a == -1)
            return num;
        swap(s[a], s[b]);
        return stoi(s);
    }
};
